libfreerdp/crypto/tls.c

```diff
@@ -159,6 +159,11 @@ static int bio_rdp_tls_read(BIO* bio, char* buf, int size)
 				break;
 
 			case SSL_ERROR_SYSCALL:
+                if (WSAGetLastError() == WSAEWOULDBLOCK)
+                {
+                    BIO_set_flags(bio, BIO_FLAGS_READ | BIO_FLAGS_SHOULD_RETRY);
+                    break;
+                }
```

Comment:
That's the part of this pull request which is most subject to discussion. I haven't debugged this myself, but I did look over the problem with Denis, and it turns out that sometimes SSL_read returns with SSL_ERROR_SYSCALL and errno set to EAGAIN/EWOULDBLOCK. We haven't been able to pinpoint the reason of this SSL_ERROR_SYSCALL, and it everything looks good in the RDP8 gateway BIO. I thought if we limit this to just WSAEWOULDBLOCK it wouldn't cause the previous problems that were caused AFAIK by ignoring all SSL_ERROR_SYSCALL, including cases where it is a real error (WSAWOULDBLOCK isn't really an error, just don't know why it ends up there though).